File: tower-defense/src/core/EventBus.js
const listeners = new Map();
export const EventBus = {
  on(event, fn) {
    if (!listeners.has(event)) listeners.set(event, new Set());
    listeners.get(event).add(fn);
    return () => this.off(event, fn);
  },
  off(event, fn) {
    listeners.get(event)?.delete(fn);
  },
  emit(event, payload) {
    listeners.get(event)?.forEach((fn) => {
      try { fn(payload); } catch (e) { console.error(e); }
    });
  },
};
export default EventBus;

File: space-exploration-procedural/nothink_3.8/src/core/EventBus.cpp
#include "EventBus.hpp"

EventBus& EventBus::instance() { static EventBus e; return e; }

unsigned long EventBus::on(const char* event, Handler h) {
    auto& v = m[std::string(event)];
    v.push_back({nextId++, std::move(h)});
    return v.back().id;
}

void EventBus::emit(const char* event) {
    auto it = m.find(std::string(event));
    if (it == m.end()) return;
    auto v = it->second; // copy: handlers may subscribe/unsubscribe
    for (auto& s : v) s.h();
}

void EventBus::off(const char* event, unsigned long token) {
    auto it = m.find(std::string(event));
    if (it == m.end()) return;
    auto& v = it->second;
    v.erase(std::remove_if(v.begin(), v.end(),
        [&](const Slot& s) { return s.id == token; }), v.end());
}
